gcloud-java-dns/src/main/java/com/google/gcloud/dns/Zone.java

```diff
@@ -0,0 +1,246 @@
+      Dns.ChangeRequestOption... options) {
+    checkNameOrIdNotNull();
+    checkNotNull(changeRequest);
+    ChangeRequest updated = null;
+    if (zoneInfo.id() != null) {
+      updated = dns.applyChangeRequest(zoneInfo.id(), changeRequest, options);
+    }
+    if (updated == null && zoneInfo.name() != null) {
+      // zone was not found by id or id is not set at all
+      updated = dns.applyChangeRequest(zoneInfo.name(), changeRequest, options);
+    }
+    return updated;
+  }
+
+  /**
+   * Retrieves an updated information about a change request previously submitted to be applied to
+   * this zone. First searches for the zone by ID and if not found then by name. Returns a {@link
+   * ChangeRequest} if found and {@code null} is the zone or the change request was not found.
+   *
+   * @param options optional restriction on what fields of {@link ChangeRequest} should be returned
+   * @return updated ChangeRequest
+   * @throws DnsException upon failure or if the zone is not found
+   * @throws NullPointerException if both zone ID and name are not initialized
+   * @throws NullPointerException if the change request does not have initialized id
+   */
+  public ChangeRequest getChangeRequest(String changeRequestId,
+      Dns.ChangeRequestOption... options) {
+    checkNameOrIdNotNull();
+    checkNotNull(changeRequestId);
+    ChangeRequest updated = null;
+    if (zoneInfo.id() != null) {
+      updated = dns.getChangeRequest(zoneInfo.id(), changeRequestId, options);
+    }
+    if (updated == null && zoneInfo.name() != null) {
+      // zone was not found by id or id is not set at all
+      updated = dns.getChangeRequest(zoneInfo.name(), changeRequestId, options);
+    }
+    return updated;
+  }
+
+  /**
+   * Retrieves all change requests for this zone. First searches for the zone by ID and if not found
+   * then by name. Returns a page of {@link ChangeRequest}s or {@code null} if the zone is not
+   * found.
+   *
+   * @param options optional restriction on listing and fields to be returned
+   * @return a page of change requests
+   * @throws DnsException upon failure or if the zone is not found
+   * @throws NullPointerException if both zone ID and name are not initialized
+   */
+  public Page<ChangeRequest> listChangeRequests(Dns.ChangeRequestListOption... options) {
+    checkNameOrIdNotNull();
+    Page<ChangeRequest> changeRequests = null;
+    if (zoneInfo.id() != null) {
+      changeRequests = dns.listChangeRequests(zoneInfo.id(), options);
+    }
+    if (changeRequests == null && zoneInfo.name() != null) {
+      // zone was not found by id or id is not set at all
+      changeRequests = dns.listChangeRequests(zoneInfo.name(), options);
+    }
+    return changeRequests;
+  }
+
+  /**
+   * Check that at least one of name and ID are initialized and throw and exception if not.
+   */
+  private void checkNameOrIdNotNull() {
+    checkArgument(zoneInfo != null && (zoneInfo.id() != null || zoneInfo.name() != null),
```

Comment:
See my [comment](https://github.com/GoogleCloudPlatform/gcloud-java/issues/610#issuecomment-178940712) in issue #610. We should check that we always have get name (builder can force it in compile time), and validate that
it is not null when creating the ZoneInfo. In that case, this check is not needed.

Also,  as discussed verbally, at that point we should change Zone methods to be based on name only.